File: app/src/main/java/com/tutv/android/ui/series_carrousel/SeriesCarrouselPresenter.java
package com.tutv.android.ui.series_carrousel;

import androidx.collection.ArraySet;

import com.tutv.android.domain.Series;
import com.tutv.android.repository.SeriesRepository;
import com.tutv.android.ui.series_carrousel.tv_banner.TvBannerView;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class SeriesCarrouselPresenter {
    private final WeakReference<SeriesCarrouselView> view;
    private final SeriesRepository seriesRepository;
    private List<Series> seriesList;
    private final CompositeDisposable disposables;

    public SeriesCarrouselPresenter(SeriesCarrouselView view, SeriesRepository seriesRepository) {
        this.view = new WeakReference<>(view);
        this.seriesRepository = seriesRepository;

        disposables = new CompositeDisposable();

        seriesList = new ArrayList<>();
    }

    private void seriesLoadSuccessful(List<Series> series) {
        this.seriesList = series;
        if(view.get() != null)
            view.get().setLoadingStatus(false);
    }

    private void seriesLoadError(Throwable e) {
    }

    public int getItemCount() {
        return seriesList.size();
    }

    public void onBindRepositoryRowViewAtPosition(int position, TvBannerView tvBannerView) {
        Series series = seriesList.get(position);

        tvBannerView.getPresenter().bind(series);
    }

    public void onViewAttached() {
        view.get().setLoadingStatus(true);

        Disposable disposable =
            seriesRepository.getFeatured()
                    .subscribeOn(Schedulers.computation())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(this::seriesLoadSuccessful, this::seriesLoadError);

        disposables.add(disposable);
    }

    public void onViewDetached() {
        disposables.dispose();
    }
}

File: app/src/main/java/com/tutv/android/ui/series_carrousel/SeriesCarrouselPresenter.kt
package com.tutv.android.ui.series_carrousel

import com.tutv.android.domain.Series
import com.tutv.android.repository.SeriesRepository
import com.tutv.android.ui.series_carrousel.tv_banner.TvBannerView
import com.tutv.android.utils.schedulers.BaseSchedulerProvider
import io.reactivex.disposables.CompositeDisposable
import java.lang.ref.WeakReference
import java.util.*

class SeriesCarrouselPresenter(
        view: SeriesCarrouselView,
        var seriesRepository: SeriesRepository,
        var schedulerProvider: BaseSchedulerProvider) {

    private val view: WeakReference<SeriesCarrouselView?> = WeakReference(view)
    private var seriesList: List<Series?> = ArrayList()
    private val disposables: CompositeDisposable = CompositeDisposable()

    private fun seriesLoadSuccessful(series: List<Series?>) {
        seriesList = series
        view.get()?.setLoadingStatus(false)
    }

    private fun seriesLoadError(e: Throwable?) {}

    fun getItemCount(): Int {
        return seriesList.size
    }

    fun onBindRepositoryRowViewAtPosition(position: Int, tvBannerView: TvBannerView?) {
        val series = seriesList[position]
        tvBannerView?.getPresenter()?.bind(series)
    }

    fun onViewAttached() {
        view.get()?.setLoadingStatus(true)

        val disposable = seriesRepository.featured
                .subscribeOn(schedulerProvider.computation())
                .observeOn(schedulerProvider.ui())
                .subscribe({ series: List<Series?> -> seriesLoadSuccessful(series) }) { e: Throwable? -> seriesLoadError(e) }

        disposables.add(disposable)
    }

    fun onViewDetached() {
        disposables.dispose()
    }

    fun onViewHolderDetached(tvBannerView: TvBannerView?) {
        tvBannerView?.onDetach()
    }

}